### PR TITLE
Refactor the deep tile matmul config and skip the single-iteration loop generation

### PR DESCRIPTION
**Pull request by zhczhong**
Track: https://github.com/intel/graph-compiler/issues/288

1. The extra memref.copy is caused by write-after-write conflict because the canonicalization pass eliminates the single-iteration loop but preserves the extracted slice and insert slice. So here skip generating them when they are single-iteration loop.
2. Refactor the config. Infer dimType according to the contractionOpInterface.
3. introduce padding cost, which minimizes the cost on padding and use divisible block if possible

### CODE REVIEW

File: lib/gc/Analysis/MatmulConfigAnalysis.cpp [hunk 1]
```diff
@@ -278,6 +281,13 @@ prepareConfigCandidates(Operation *root, CPUTargetDescriptionAnalysis &sysDesc,
           : getCandidate((uint32_t)shape[2],
                          shape[2] >= noSmallBlockNeedThreshold ? 8U : 1U, 256U);
 
+  if (allowIndivisibleInnerblock) {
+    innerMostKBlockCandidates = {16, 32, 64};
+    innerMostNBlockCandidates = {16, 32, 64};
+    NBlockCandidates = innerMostNBlockCandidates;
+    KBlockCandidates = innerMostKBlockCandidates;
```
Comment (yifeizh2) [inline lib/gc/Analysis/MatmulConfigAnalysis.cpp:288]: So after the change here, innermost Kblock will only be one of 16/32/64 if `allowIndivisibleInnerblock` is true?

Comment (zhczhong) [reply lib/gc/Analysis/MatmulConfigAnalysis.cpp:288]: Yes, your understanding is correct

File: lib/gc/Transforms/DeepTileContractionOp.cpp [hunk 2]
```diff
@@ -107,6 +107,16 @@ tensorViewRankedTensor(RewriterBase &rewriter, RankedTensorType outTensorType,
   return result;
 }
 
+// Check if the loop is dummy loop(has only one iteration)
+bool isDummyLoop(LoopLikeOpInterface loop) {
```
Comment (ciyongch) [inline lib/gc/Transforms/DeepTileContractionOp.cpp:111]: Seems this is another impl besides "promoteIfSingleIteration()"?

File: lib/gc/Transforms/DeepTileContractionOp.cpp [hunk 3]
```diff
@@ -260,31 +270,34 @@ static void getMatmulParallelDims(linalg::LinalgOp linalgOp,
...
-      Value viewResult = tensorViewRankedTensor(
-          rewriter, targetTensor, newExtractSliceOp->getResult(0));
-      rewriter.replaceOp(extractSlice, viewResult);
-    } else {
-      rewriter.replaceOp(extractSlice, newExtractSliceOp);
-    }
+    rewriter.replaceOp(op, currentOp);
+  }
+
+  if (shrinDimNum > 0) {
```
Comment (ciyongch) [inline lib/gc/Transforms/DeepTileContractionOp.cpp:291]: nit: shrinDimNum->shrinkDimNum?

File: lib/gc/Transforms/DeepTileContractionOp.cpp (hunk 2 above)
Comment (zhczhong) [reply lib/gc/Transforms/DeepTileContractionOp.cpp:111]: Yes, the purpose is similar but the `promoteIfSingleIteration()` erase the single iteration loop only after the loop is created. The extract/insert slices will be preserved, which causes the WAW conflict in bufferization and introduces more `memref.copy`. Here skip generating the single iteration loop and its corresponding extract/insert_slice op in the creation  phase once detecting the single iteration.

File: lib/gc/Analysis/MatmulConfigAnalysis.cpp [hunk 4]
```diff
@@ -55,9 +55,10 @@ bool validateConfig(const MatmulConfig &cfg) {
 std::vector<uint32_t>
 getCandidate(uint32_t num, uint32_t floor,
              uint32_t ceil = std::numeric_limits<uint32_t>::max()) {
+  int defaultBlock = 32;
   // factor
   std::vector<uint32_t> candidates;
-  uint32_t upperbound = std::min(num, ceil);
+  uint32_t upperbound = std::min(llvm::divideCeil(num, defaultBlock), ceil);
```
[the remark this answers is not in the document]
Comment (zhczhong) [reply lib/gc/Analysis/MatmulConfigAnalysis.cpp]: Thanks for pointing it out! should be `llvm::divideCeil(num, defaultBlock) * defaultBlock`

File: lib/gc/Transforms/DeepTileContractionOp.cpp (hunk 3 above)
Comment (zhczhong) [reply lib/gc/Transforms/DeepTileContractionOp.cpp]: Fixed

File: include/gc/Analysis/MatmulConfigAnalysis.h [hunk 5]
```diff
@@ -47,15 +47,67 @@ inline SmallVector<unsigned> extractDimTypeIdx(ArrayRef<DimType> tyList,
   return idxList;
 }
 
+inline void getDimTypeFromIterators(linalg::LinalgOp linalgOp,
+                                    SmallVectorImpl<DimType> &dimTypes) {
+  SmallVector<utils::IteratorType> iteratorTypes =
```
Comment (yifeizh2) [inline include/gc/Analysis/MatmulConfigAnalysis.h:52]: Can you explicitly specify `mlir::utils::IteratorType` here?

File: lib/gc/Transforms/DeepTileContractionOp.cpp (hunk 2 above)
Comment (ciyongch) [reply lib/gc/Transforms/DeepTileContractionOp.cpp:111]: I see, thanks for the explanation.

File: include/gc/Analysis/MatmulConfigAnalysis.h (hunk 5 above)
Comment (zhczhong) [reply include/gc/Analysis/MatmulConfigAnalysis.h]: add `mlir` namespace here

**Comment by ciyongch** [comment]
Overall LGTM, do you think we should add a single simple case to cover the case of skipping single-iteration loop generation?

**Comment by zhczhong** [comment]
> Overall LGTM, do you think we should add a single simple case to cover the case of skipping single-iteration loop generation?

This test could cover this case. The file check will check two `scf.forall` before this PR but the single-iteration `scf.forall` will be skipped now.

https://github.com/intel/graph-compiler/blob/613a333ee5e14731feeca40b54c57adb07ca7edc/test/mlir/test/gc/Transforms/deepTileContractionNamedOp.mlir#L127

**Comment by ciyongch** [comment]
Please help to rebase the code base, then we can merge it.

**Comment by zhczhong** [comment]
> Please help to rebase the code base, then we can merge it.

The code has been rebased now

**Review by ciyongch — APPROVED** [review]

**Review by yifeizh2 — APPROVED** [review]

**Pull request merged**